ember-client/shared/src/main/scala/org/http4s/ember/client/internal/EmberWSClient.scala

```diff
@@ -0,0 +1,126 @@
+import scodec.bits.ByteVector
+
+import java.util.Base64
+
+private[client] object EmberWSClient {
+  def apply[F[_]](
+      emberClient: Client[F]
+  )(implicit F: Async[F]): F[WSClient[F]] =
+    SecureRandom.javaSecuritySecureRandom[F].map { random =>
+      WSClient[F](respondToPings = false) { wsRequest =>
+        for {
+          randomByteArray <- Resource.eval(random.nextBytes(16))
+
+          uriScheme = wsRequest.uri.scheme.map(scheme =>
+            scheme.value match {
+              case "wss" => Uri.Scheme.https
+              case "ws" => Uri.Scheme.http
+              case _ => scheme
+            }
+          )
+
+          httpWSRequest = Request[F]()
+            .withUri(wsRequest.uri.copy(uriScheme))
+            .withHeaders(
+              Headers(
+                upgradeWebSocket,
+                connectionUpgrade,
+                supportedWebSocketVersionHeader,
+                new `Sec-WebSocket-Key`(ByteVector(Base64.getEncoder().encode(randomByteArray))),
+              )
+            )
+            .withMethod(Method.GET)
+
+          socketOption <- getSocket(emberClient, httpWSRequest)
+          socket <- socketOption.liftTo[F](new RuntimeException("Not an Ember client")).toResource
+
+          closeFrameDeffered <- F.deferred[WebSocketFrame.Close].toResource
+
+          clientReceiveQueue <- Queue.bounded[F, WebSocketFrame](100).toResource
+          clientSendChannel <- Channel.bounded[F, WebSocketFrame](100).toResource
+
+          _ <- socket.reads
+            .through(decodeFrames(true))
+            .foreach {
+              case f @ WebSocketFrame.Close(_) =>
+                closeFrameDeffered.complete(f).ifM(clientReceiveQueue.offer(f), F.unit)
+              case f =>
+                closeFrameDeffered.tryGet.flatMap { x =>
+                  if (x.isDefined) F.unit else clientReceiveQueue.offer(f)
+                }
+            }
+            .compile
+            .drain
+            .background
+
+          sendingFinished <- clientSendChannel.stream
+            .foreach(f => frameToBytes(f, true).traverse_(c => socket.write(c)))
+            .compile
+            .drain
+            .background
+
+          _ <- Resource.onFinalize {
+            MonadThrow[F]
+              .fromEither(WebSocketFrame.Close(1000, "Connection automatically closed"))
+              .flatMap(clientSendChannel.send(_))
+              .void *> clientSendChannel.close.void *> sendingFinished.void
+          }
+        } yield new WSConnection[F] {
+          def receive: F[Option[WSFrame]] = clientReceiveQueue.take.map {
+            case f @ WebSocketFrame.Close(_) =>
+              clientSendChannel.close.void
+              toWSFrame(f).some
+            case f =>
+              toWSFrame(f).some
+          }
+          def send(wsf: WSFrame): F[Unit] =
+            toWebSocketFrame(wsf).flatMap {
+              case WebSocketFrame.Close(_) =>
+                clientSendChannel.close.void
```

Comment:
Hmm, are we forgetting to send the close frame here? In any case, we can use the new API `channel.closeWithElement(...)` to guarantee that the close frame is the last frame that is sent.

Also I don't think we should `void`, if the `Channel` is closed we should probably raise an error for trying to send a frame a on a closed connection.